lib/input_sanitizer/restricted_hash.rb

```diff
@@ -1,8 +1,33 @@
 module InputSanitizer
   class RestrictedHash < Hash
     def initialize(allowed_keys)
-      @allowed_keys = allowed_keys
-      super() { |hash, key| default_for_key(key) }
+      @allowed_keys = Set.new(allowed_keys)
+      super()
+    end
+
+    def [](key)
+      raise_not_allowed(key) unless key_allowed?(key)
+      fetch(key, nil)
+    end
+
+    def []=(key, val)
+      @allowed_keys.add(key)
```

Comment:
I'm just passing by this PR, sorry if it's an out-of-context question. 
Does it make sense to add allowed keys when someone writes to a key?  It seems that the previous behavior had a whitelist of keys on hash initialization.

Reply:
What do you mean by "had a whitelist of keys on hash initialisation" ?
Previous implementation was setting `allowed keys` on initialisation and so is this one (with a difference that current `allowed keys` is a set instead of whatever/array to prevent duplicates ((they weren't bad tbh but still shouldn't be allowed))) 

Previous implementation was working with new keys added to hash because of the flow of [] method in ruby hash implementation. 
eg 
```rhash = RestrictedHash.new([:a])
rhash[:b] = 1
``` and then ```
rhash.instance_variable_get(:@allowed_keys)
=> [:a] 
rhash[:a]
=> nil # becasue default proc says so 
rhash[:b]
=> 1 #because hash [] first checks if there is a value for that key. Only if there's no value, default_proc is called
```

Reply:
thanks for this clear explanation